src/NHibernate/Engine/Query/QueryPlanCache.cs

```diff
@@ -58,15 +58,38 @@ public IQueryExpressionPlan GetHQLQueryPlan(IQueryExpression queryExpression, bo
 				{
 					log.Debug("unable to locate HQL query plan in cache; generating ({0})", queryExpression.Key);
 				}
+
+				var refinableQuery = queryExpression as IRefinableKeyQueryExpression;
+				var wasAlreadyRefined = refinableQuery?.RefinedKey;
+
 				plan = new QueryExpressionPlan(queryExpression, shallow, enabledFilters, factory);
 				planCache.Put(key, plan);
+
+				if (wasAlreadyRefined == false && refinableQuery.RefinedKey)
```

Comment:
Here, instead of caching with original and refined key, we could take quite a different path: cache nothing. So the `RefinedKey` would be renamed `NotCacheable`, it would be checked right after plan generation, and if `true`, the plan would not be put in cache. It would have several advantages:

- Simpler.
- Would solve last #1363 trouble without the workaround.
- Would avoid retaining captured object instances.
- Would avoid bloating the cache with as many entries as constant values used with queries having non-parameterized constants.

In #1363, some users were even asking for a way to fully disable the query plan cache, writing they were not feeling a performance hit for not having it. Here it would disable it only for queries embedding constants which are not translated to HQL parameters, so it will be even less likely to have a negative impact.